### PR TITLE
Adds Trinket Compare

### PR DESCRIPTION
**Pull request by Ellimistdev**
Needed a good way to see a trinket's impact across multiple specs.

### Adds: 
- Top level nav "Trinkets"
- New Trinket submenu to swap between item, item level, and fight styles

trinkets_balefire_branch_454_castingpatchwerk
trinkets_caged_horror_486_castingpatchwerk3

### Outstanding todos:
MVP:
~~- Hook into db instead of fetching everything like a goblin~~
~~- Need to add trinket-specific queries instead of needing to get ALL trinkets for ALL specs 😒~~
~~- Server-side data handling~~

Midterm:

- [x] Better trinket selection dropdown (scrolling at a minimum)
- [x] Item_level handling for non-existent item levels for items
- [x] Update fight_style in chart title.
- [x] Trinket menu colors? ~~Just picked Rogue.~~ is now priest

Longterm:
- Swap item_name for item_id throughout to aid in localization
- Drilldown links: trinket_compare to trinkets

### CODE REVIEW

**Comment by Bloodmallet** [comment]
This looks great! Let me add some information:
- the db does not "know" about simulation results and their inner workings. it is aware of where their jsons exist. therefore fetching all simulation results "like a goblin" is the way to go
- server side tries to be as uninvolved with game data as possible. it tries to let `bloodytools` in cooperation with `simc_support` do their thing and just pipe the result through to the frontend js

Midterm:

    Better trinket selection dropdown (scrolling at a minimum): [this](https://getbootstrap.com/docs/4.0/components/dropdowns/#options)  should hopefully help
    Item_level handling for non-existent item levels for items: when collecting trinket information each trinket should already know about which itemlevels are present for it. this can be used to filter the list or create it
    Update fight_style in chart title.: 👍 
    Trinket menu colors? Just picked Rogue.: I'd probably pick priest because it's the most neutral "color"

Longterm:

    Swap item_name for item_id throughout to aid in localization: 👍 
    Drilldown links: trinket_compare to trinkets: sounds great!

File: general_website/views/general.py
Comment (Bloodmallet) [inline general_website/views/general.py:1]: I'd put all this logic into the chart importing javascript code instead of creating endpoints that basically self-crawl the website.

**Comment by Ellimistdev** [comment]
Finally picked this back up! 

I reworked this so that it does not constantly self-crawl the site. Given that we currently receive all item data for any one spec, to generate these charts, we only need to hit each spec once and cache the result in localstorage (updated every 30 minutes). Additionally, I extracted several functions from `bloodmallet_chart_import()` to reduce complexity and improve reusability.

As far as I can tell, these changes don't impact existing functionality, though I have not ventured into accounts and custom charts.

TrinketCompare

**Comment by Bloodmallet** [comment]
Sounds great! I'll need some time to review it. I'm currently busy preparing my own chart implementations and that's occupying my attention. I'll hopefully get back to you soon on this. 👍

**Comment by Ellimistdev** [comment]
Updated for 11.1. 

Additional improvements: Added bm-utils for shared helpers, improved error handling, localization, caching, populating trinket compare menu with available items & item levels.

TrinketCompare11 1

**Comment by Ellimistdev** [comment]
@Bloodmallet ready for review :)

File: general_website/migrations/0008_alter_simulation_custom_apl_and_more.py
Comment (Ellimistdev) [inline general_website/migrations/0008_alter_simulation_custom_apl_and_more.py:1]: Added automatically when following the development setup `python manage.py migrate`

**Comment by Bloodmallet** [comment]
I'll try to check this out in more depth this coming weekend. They idea is still great!

Btw from a glance at your provided screenshots (thanks btw for those) was an itemlevel selection a specific consideration? Otherwise similar to the normal charts each trinket could straight up show all itemlevels. So the user would only need to select the trinket.

**Comment by Ellimistdev** [comment]
> I'll try to check this out in more depth this coming weekend. They idea is still great!
> 
> Btw from a glance at your provided screenshots (thanks btw for those) was an itemlevel selection a specific consideration? Otherwise similar to the normal charts each trinket could straight up show all itemlevels. So the user would only need to select the trinket.

That's a great question. Initially, looking at a single item level across specs seemed like a smaller piece of work, which informed the data structure I'm using. 

image

Which is pretty different from the structure used by the class-specific trinket charts

image

Ultimately, showing all iLvls simultaneously is a better user experience, but it wouldn't be a quick refactor.

File: general_website/static/general_website/js/bm-charts.js [hunk 1]
```diff
@@ -589,18 +818,11 @@ class BmChartData {
...
-            "cn_CN": "cn",
-            "de_DE": "de",
-            "es_ES": "es",
-            "fr_FR": "fr",
-            "it_IT": "it",
-            "ko_KR": "ko",
-            "pt_BR": "pt",
-            "ru_RU": "ru"
-        };
+        const subdomain = window.bmUtils.wowheadSubdomains;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js]: I see why you did this and it absolutely makes sense to use another file to ensure not doubling the information. But this importer is [intended as a file other sites can import](https://bloodmallet.com/faq#data-usage) to show the charts on their own, relying on other files becomes a hassle. That's why I believe this can't get moved out.

File: general_website/static/general_website/js/bm-charts.js [hunk 2]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
+        let chart_type;
+        let fight_style;
+        let item_name;
+        let item_level;
+
         if (chart_anchor.dataset.hasOwnProperty("chartId")) {
+            console.debug(`Chart has chartId:`, chart_anchor.dataset.chartId);
             // if chart_id -> load id
-            let chart_id = chart_anchor.dataset["chartId"];
+            let chart_id = chart_anchor.dataset?.chartId;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js:1885]: Why would it be better here to basically create an invalid request with an `undefined` chart id instead of crashing the function?

File: general_website/static/general_website/js/bm-charts.js [hunk 3]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
             // console.log("Identified chart id:", chart_id);
             request_endpoint = endpoint + "/" + chart_id;
-        } else if (chart_anchor.dataset.hasOwnProperty("wowClass") && chart_anchor.dataset.hasOwnProperty("wowSpec")) {
-            // elif class & spec [& chart type] [& fight style] -> load general data
-            let wow_class = chart_anchor.dataset["wowClass"];
-            let wow_spec = chart_anchor.dataset["wowSpec"];
-            let chart_type = "trinkets";
+        } else if ("wowClass" in chart_anchor.dataset && "wowSpec" in chart_anchor.dataset &&
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js:1888]: While I like this notation more I thought in js `in` wasn't as supported yet for this kind of checks?

File: general_website/static/general_website/js/bm-charts.js [hunk 4]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
+            chart_anchor.dataset.wowClass && chart_anchor.dataset.wowSpec) {
+            let wow_class = chart_anchor.dataset?.wowClass;
+            let wow_spec = chart_anchor.dataset?.wowSpec;
+            chart_type = "trinkets";
+            fight_style = "castingpatchwerk";
+
             if (chart_anchor.dataset.hasOwnProperty("type")) {
-                chart_type = chart_anchor.dataset["type"];
+                chart_type = chart_anchor.dataset?.type;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js]: I'm not familiar enough with js, so please bear with my questions, but what is this change improving?

File: general_website/static/general_website/js/bm-charts.js [hunk 5]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
             if (chart_anchor.dataset.hasOwnProperty("fightStyle")) {
-                fight_style = chart_anchor.dataset["fightStyle"];
+                fight_style = chart_anchor.dataset?.fightStyle;
             }
+
             // console.log("Identified chart_import for standard", chart_type, "chart of fight_style", fight_style, "for", wow_spec, wow_class);
             request_endpoint = [endpoint, chart_type, fight_style, wow_class, wow_spec].join("/");
+        } else if ("type" in chart_anchor.dataset && chart_anchor.dataset.type === "trinket_compare") {
+            // Handle trinket_compare
+            item_name = chart_anchor.dataset?.itemName || "aberrant_spellforge";
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js]: Instead of hardcoding a fallback name here get the first trinket of the name-sorted list.

File: general_website/static/general_website/js/bm-charts.js [hunk 6]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
+            item_level = chart_anchor.dataset?.itemLevel || "636";
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js]: Dynmically get the highest itemlevel as fallback.

File: general_website/static/general_website/js/bm-charts.js [hunk 7]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
+            chart_type = chart_anchor.dataset?.type;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js]: Haven't you already established in the `if` that this key exists?

File: general_website/static/general_website/js/bm-charts.js [hunk 8]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
+            fight_style = chart_anchor.dataset?.fightStyle || "castingpatchwerk";
+            request_endpoint = [endpoint, chart_type, fight_style, item_name, item_level].join("/");
         }
         // console.log("bloodmallet.com: loading chart from", request_endpoint);
+        try {
+            let data;
+            if (chart_type === "trinket_compare") {
+                data = await getTrinketDataAsync(item_name, item_level, fight_style);
+            } else {
+                const response = await fetch(request_endpoint);
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js:1911]: Why did you refactor the `XMLHttpRequest` to a `fetch` call?

File: general_website/static/general_website/js/bm-charts.js [hunk 9]
```diff
@@ -1611,79 +1839,101 @@ function bm_import_charts() {
...
-                    // console.log(bm_data.data_type);
-                    // get chart type from loaded data
-                    let chart = BmBarChart;
-                    if (bm_data.data_type === "secondary_distributions") {
-                        // create Chart based on chart type 
-                        chart = BmRadarChart;
-                    }
+            chart_anchor.dataset.loadedData = JSON.stringify(data);
+            let bm_data = new BmChartData(chart_anchor);
+            let chart = bm_data.data_type === "secondary_distributions" ? BmRadarChart : BmBarChart;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js]: While I enjoy ternaries I wrote this withan if-conditional expecting more chart types to arrive in the future. While there are no immediate plans I thought it would be clearer which is loaded using such blocks instead of an ever growing long line of ternaries.

File: general_website/static/general_website/js/bm-charts.js [hunk 10]
```diff
@@ -3,6 +3,192 @@ const BmChartStyleUrl = "/static/general_website/css/bm-charts.css";
 const BmTooltipJsId = "bm-tooltip-javascript";
 const BmTooltipJsUrl = "/static/general_website/js/bm-tooltips.js";
 
+let trinketDataCache = {};
+const TRINKET_DATA_CACHE_KEY = 'trinketData';
+const TRINKET_DATA_CACHE_EXPIRY = 30 * 60 * 1000;  // 30 minutes in milliseconds
+
+const isTrinketDataCacheValid = () => {
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-charts.js:10]: What's the benefit of this function notation compared to the already present notation in the file?

Yours
```js
const functionName = (param1, param2) => {}
```

versus already present
```js
function functionName(param1, param2) {}
```

File: general_website/static/general_website/js/bm-utils.js [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    return languageMap[langCode] || langCode;
+};
+
+/**
+ * Detects the user's language from element, cookies or browser settings
+ * @param {HTMLElement} element Optional element to check for language attribute
+ * @returns {string} The full language code (e.g., "en_US")
+ */
+const detectUserLanguage = (element = null) => {
+    let langCode = getLanguageFromCookie() || getLanguageFromDataset(element) || getLanguageFromBrowser();
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-utils.js]: If the website decided to set a language on the html element I believe this should have precedence over everything.

1. html element (dataset)
2. cookie
3. browser

File: general_website/static/general_website/js/bm-utils.js [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+ * @param {string} text The text to format
+ * @param {string} type The type of formatting to apply (e.g., "slug", "fight_style")
+ * @param {Object} formatDictionary Optional dictionary for special formatting like fight styles
+ * @returns {string} The formatted text
+ */
+const formatText = (text, type) => {
+    if (!text) return "Loading...";
+
+    switch (type) {
+        case "slug":
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/bm-utils.js]: Is there a way to create an Enum with these options, that is also available when working on other files in VS Code?

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+// Utility Functions
+// ==========================
+  
+  /**
+   * Create a default state object with optional overrides
+   */
+  const createInitialState = (overrides = {}) => ({
+    data_type: 'trinket_compare',
+    fight_style: 'castingpatchwerk',
+    wow_class: 'priest',
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js:16]: Why does this need a wow class at all?

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+  // ==========================
+  // Data Fetching and Processing
+  // ==========================
+  
+  /**
+   * Fetch available trinket data
+   */
+  const fetchAvailableTrinkets = async (fightStyle) => {
+    try {
+      const fetchFunction = window.fetchAndProcessDataAsync ?? fetchAndProcessDataAsync;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js:45]: What happens here?

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 15]
```diff
@@ -0,0 +1,459 @@
...
+      const data = await fetchFunction?.(fightStyle);
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js:46]: It gets wilder and wilder for me. I thought `?.` allowed access to an object and in case this object doesn't have the follow up key it'll automatically return `undefined`. Now this calls...a function? Is this some sort of implicit error handling? So if the call fails/the function call throws an error, this instread will produce `undefined` again?

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 16]
```diff
@@ -0,0 +1,459 @@
...
+      return processTrinketsFromData(data);
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js:52]: In case the above assumption is correct I believe the code should already inform the user here about the missing data. Cause we're encountering the problem here. We expect data but don't get any. If we use `undefined` to create a chart it's suddenly up to the user to figure out where the weird chart creation call came from without data. At least to me that would be a suboptimal debugging experience.

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 17]
```diff
@@ -0,0 +1,459 @@
...
+      console.warn("No trinket data available to process");
+      return availableTrinkets;
+    }
+  
+    try {
+      // Detect user language
+      const userLanguage = window.bmUtils?.detectUserLanguage() || 'en_US';
+  
+      for (const trinketKey in data.items) {
+        if (trinketKey === "baseline") continue;
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js]: I know this works. I know some people love to not use parenthesis for such cases. I don't. I like to be explicit about what is the intended part of the conditional. And In case this section grows the diff will only add elements within the already prepared block and not "alter" the if-condition line.

```suggestion
        if (trinketKey === "baseline") {
          continue;
        }
```

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+  
+        let trinketName = trinketKey.replace(/_/g, ' ')
+          .replace(/\b\w/g, c => c.toUpperCase()); // Default formatting
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js]: Wait...didn't you add a formating fucntion for such a case?

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+          }
+        }
+  
+        availableTrinkets.push({
+          key: trinketKey,
+          name: trinketName
+        });
+      }
+  
+      return availableTrinkets.sort((a, b) => a.name.localeCompare(b.name));
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js:95]: Oh nice. I didn't know about localCompare.

File: general_website/static/general_website/js/navbar_trinket_menu.js [hunk 20]
```diff
@@ -0,0 +1,459 @@
...
+        item_levels: chartData.item_levels || [],
+        available_trinkets: availableTrinkets
+      });
+  
+      await update_navbarTrinketMenu(state);
+  
+      console.debug(`Loaded ${availableTrinkets.length} available trinkets`);
+  
+      // Set up observer to watch for future changes
+      const observer = new MutationObserver(async (mutations) => {
```
Comment (Bloodmallet) [inline general_website/static/general_website/js/navbar_trinket_menu.js:388]: I need to read up on `MutationObserver`

**Review by Bloodmallet — COMMENTED** [review]
This. Is. Awesome!
Thank you very much for this PR. This looks like a lot of work and care was put into it. I added a few comments to better understand your intentions and what's going on. Please have a look.

I love there are several opportunities for me to learn a lot.

I'm very much looking forward to get this merged soon.

**Comment by Bloodmallet** [comment]
I want to add that after reading your PR I want to look into applying the following constructs I've seen in your PR to the whole code base:

- localStorage
- MutationObserver
- fetch

```js
let a = get_something || default;
```

```js
let a = object?.magicalKey;
```

Registering functions (and classes?) to the window
```js
window.customkey = {
    function1,
    function2,
    class1
};
```

Your `bmUtils.createElement` versus creating all elements "by hand" as is currently done everywhere else.

Thank you very much

File: general_website/static/general_website/js/bm-charts.js (hunk 4 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/bm-charts.js]: Good question! This pattern is called Optional Chaining. The original implementation `dataset["type"]` assumes that `chart_anchor.dataset` definitely exists, and will throw a runtime error if it doesn't. While `dataset?.type` will return undefined instead of crashing if `dataset` doesn't exist. This allows the code to safely execute if the data is available but the DOM isn't fully rendered - which is helpful in async execution. In fact, I should have removed the preceding line as well `if (chart_anchor.dataset.hasOwnProperty("type"))`, since that's implicit in optional chaining.
TL;DR: Cleaner and safer.

File: general_website/static/general_website/js/bm-charts.js (hunk 3 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/bm-charts.js:1888]: If you can believe it - `"key" in object` notation has been around since ES3 (the early 2000s)! Minor "gotcha" here is that `"prop" in obj` checks inherited props as well (which is especially good for browser code since we're just checking for the existence of the property without wanting to traverse the DOM) while `obj.hasOwnProperty("prop")` ignores inherited props.

File: general_website/static/general_website/js/bm-charts.js (hunk 2 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/bm-charts.js:1885]: This goes back to safe async execution - If the DOM is still loading and the `chartId` is missing, we can retry `bm_import_charts()` without affecting the user experience. Without this, the application will hard crash and require a reload on the user's end (which might not resolve the missing attribute). This opens us up to gracefully handle missing charts if we want to e.g. on pages with multiple charts loading.

File: general_website/static/general_website/js/bm-charts.js (hunk 7 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/bm-charts.js]: Great catch 🙏

File: general_website/static/general_website/js/bm-charts.js (hunk 8 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/bm-charts.js:1911]: Many reasons! Essentially, `fetch` is "promise-based" which allows it to support `async/await` natively while `XMLHttpRequest` is "callback-based" and requires manual handling for async requests. [More info here](https://www.greatfrontend.com/questions/quiz/what-are-the-differences-between-xmlhttprequest-and-fetch)

File: general_website/static/general_website/js/bm-charts.js (hunk 10 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/bm-charts.js:10]: Most modern frameworks prefer arrow functions over function declarations for cleaner and more consistent syntax, but there are also some major differences under the hood. Here, in particular, `function functionName(param)` does some stuff that we don't need, such as rebinding `this`, creating an `arguments` object, and hoisting the function.

File: general_website/static/general_website/js/navbar_trinket_menu.js (hunk 13 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/navbar_trinket_menu.js:16]: I suppose it doesn't 😁, this was to tap into the existing formatting elsewhere for coloring the text.

File: general_website/static/general_website/js/navbar_trinket_menu.js (hunk 14 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/navbar_trinket_menu.js:45]: This line is using the "nullish coalescing operator" (`??`) to assign the first defined value between two options. If `window.fetchAndProcessDataAsync` is defined it gets used, otherwise it falls back to the local `fetchAndProcessDataAsync` function. 
This allows users to inject their own data handler if they're so inclined and opens up the ability for us to mock this code later on for testing. Probably overkill tbh

File: general_website/static/general_website/js/navbar_trinket_menu.js (hunk 15 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/navbar_trinket_menu.js:46]: Yes and no. There's no implicit error handling here from inside the function, but it does do type checking before calling the function. 
It is equivalent to 
```js
let data;
if (typeof fetchFunction === 'function') {
  data = await fetchFunction(fightStyle);
} else {
  data = undefined;
}
```
Errors thrown inside the function will be caught in the enclosing `try/catch`

File: general_website/static/general_website/js/navbar_trinket_menu.js (hunk 16 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/navbar_trinket_menu.js:52]: Absolutely right. While `processTrinketsFromData` validates the structure and contents of the response, logs errors, and returns a safe fallback (`[]`), there is still a gap where we need to validate that `data` is non-empty before passing it forward.

File: general_website/static/general_website/js/navbar_trinket_menu.js (hunk 20 above)
Comment (Ellimistdev) [reply general_website/static/general_website/js/navbar_trinket_menu.js:388]: `MutationObserver` is so cool 😁This avoids manually calling a chart update function after each fetch and just reacts to when the data in the chart actually changes instead of looking for events or something

**Comment by Ellimistdev** [comment]
Cleaning up [my branch ](https://github.com/Ellimistdev/bloodmallet_web_frontend/tree/bmRefactor) and will push changes this week, but got the multilevel trinket display working:
multiilvl

Also did a decent amount of refactoring w.r.t. chart/component creation 
 1. Modular Architecture

Before: Everything was in a single monolithic file with global functions
After: Clear separation into classes (BmChartData, BmBarChart, BmRadarChart, BmUIUtils, BmChartComponents)
Value: Better maintainability, testability, and code organization

2. Utility Class Introduction

New: BmUIUtils class with comprehensive helper methods
Methods Added:
- Language detection (detectUserLanguage, getLanguageFromCookie, etc.)
- Text formatting (formatText, formatNumber, capitalizeFirstLetters)
- DOM creation (createElement, createDiv, createSpan)
- Tooltip system integration
- Safe JSON parsing

Value: Centralized utilities, reduced code duplication, better error handling

3. Component-Based Chart Creation

New: BmChartComponents class for chart element creation
Methods: createLegend, createAxisTitles, createChartRow, createTooltip, etc.
Value: Modular chart building, easier customization and testing

Code Quality Improvements
4. Better Documentation

Before: Minimal JSDoc comments
After: Comprehensive JSDoc with parameter types, return types, and examples
Value: Better developer experience, easier onboarding

5. Static Methods and Organization

Before: Global functions scattered throughout
After: Logical grouping in static methods within classes
Value: Better namespace management, clearer API

6. Enhanced Filter System

New: Static methods for filtering (passesActivePassiveFilter, passesItemLevelFilter, compareDPSValues)
Value: More maintainable filtering logic, easier to extend

**Comment by Bloodmallet** [comment]
Thank you for the updates. I'm scheduling this for review next weekend. Hopefully I'll be able to also merge it by then. Looking forward to it!

**Comment by Bloodmallet** [comment]
This looks incredible even without your announced other changes! So many things to learn for me in it. Thank you very much.

I'm going ahead and merge this already. You worked hard on this for such a long time. It's time to release it! Please open a new PR for the additional changes you announced above.
